lib/SPIRV/SPIRVTypeScavenger.cpp

```diff
@@ -53,156 +130,498 @@
+  if (auto *FT = dyn_cast<FunctionType>(T)) {
+    for (Type *Inner : FT->params())
+      if (hasTypeVariable(Inner, TypeVarNum))
+        return true;
+    return hasTypeVariable(FT->getReturnType(), TypeVarNum);
+  }
+  unsigned CheckNum;
+  if (isTypeVariable(T, CheckNum)) {
+    return TypeVarNum == CheckNum;
+  }
+  return false;
+}
+} // anonymous namespace
+
+Type *SPIRVTypeScavenger::substituteTypeVariables(Type *T) {
+  if (auto *TPT = dyn_cast<TypedPointerType>(T))
+    return TypedPointerType::get(substituteTypeVariables(TPT->getElementType()),
+                                 TPT->getAddressSpace());
+  if (auto *VT = dyn_cast<VectorType>(T))
+    return VectorType::get(substituteTypeVariables(VT->getElementType()),
+                           VT->getElementCount());
+  if (auto *AT = dyn_cast<ArrayType>(T))
+    return ArrayType::get(substituteTypeVariables(AT->getElementType()),
+                          AT->getNumElements());
+  if (auto *FT = dyn_cast<FunctionType>(T)) {
+    SmallVector<Type *, 4> ParamTypes;
+    for (Type *Inner : FT->params())
+      ParamTypes.push_back(substituteTypeVariables(Inner));
+    Type *ReturnTy = substituteTypeVariables(FT->getReturnType());
+    return FunctionType::get(ReturnTy, ParamTypes, FT->isVarArg());
+  }
+  unsigned TypeVarNum;
+  if (isTypeVariable(T, TypeVarNum)) {
+    TypeVarNum = UnifiedTypeVars.join(TypeVarNum, TypeVarNum);
+    Type *&SubstTy = TypeVariables[TypeVarNum];
+    // A value in TypeVariables may itself contain type variables that need to
+    // be substituted. Substitute these as well.
+    if (SubstTy)
+      return SubstTy = substituteTypeVariables(SubstTy);
+    // Even if it's not fully resolved, return the leader of the current
+    // equivalence class instead. This allows for easier scanning of recursive
+    // type declarations.
+    return TargetExtType::get(T->getContext(), "typevar", {}, {TypeVarNum});
+  }
+  return T;
+}
+
+bool SPIRVTypeScavenger::unifyType(Type *T1, Type *T2) {
+  T1 = substituteTypeVariables(T1);
+  T2 = substituteTypeVariables(T2);
+  if (T1 == T2)
+    return true;
+
+  auto SetTypeVar = [&](unsigned TypeVarNum, Type *ActualTy) {
+    // .findLeader doesn't work in uncompressed mode, so use .join with itself
+    // to find the leader.
+    unsigned Leader = UnifiedTypeVars.join(TypeVarNum, TypeVarNum);
+
+    // This method might be called with T1 as a concrete type containing
+    // pointers, and we want to make sure those don't leak into type variables.
+    // Guard against that here.
+    ActualTy = allocateTypeVariable(ActualTy);
+
+    // Check for recursion in type variables. Such recursive types generally
+    // cannot be correctly typed.
+    if (hasTypeVariable(ActualTy, Leader))
+      return false;
+
+    LLVM_DEBUG(dbgs() << "Type variable " << TypeVarNum << " is " << *ActualTy
+                      << "\n");
+    assert(!TypeVariables[Leader] && "Type was already fixed?");
```

Comment:
Did we resize `TypeVariables` somewhere? If not, doesn't it lead to our-of-bound access? I'm not sure if it will be nullptr in such case.